Edits to spatial train/val/test, additional performance metrics

jds485 opened the pull request:
- I found that validation segments were appearing in the test set, and training segments were appearing in validation. I edited some if/else statements to ensure that only test segments are in testing, and only validation are in validation. I'm not sure if the previous implementation was as-intended or a bug.
- I added bi-weekly, monthly, and yearly performance metrics, and a by-year summary of daily data. Also added log-KGE, and recommending that 21 be the minimum number of observations needed to compute a performance metric. For the 10th and 90th percentile metrics, this would give 2 samples each.
- I added args to preprocessing to filter the x_data by an earliest and/or latest date. With these arguments, the xarray of data could contain NAs/NaNs before or after these dates, respectively. I think this is useful when there are lagged attributes that would have NAs in the lag before the 1st non-NA point.

janetrbarclay reviewed (changes requested):
Grr. I intended to click "request changes" before.  Sorry for the change!!

jds485 commented on river_dl/preproc_utils.py:
```diff
@@ -632,9 +633,12 @@ def prep_y_data(
+    sites will be witheld from training and testing
     :param test_sites: [list of site_ids] sites to retain for testing. These
     sites will be witheld from training and validation
+    :param strict_spatial_partition: [bool] when True, the test set does not 
```
How about this description? 
"when True and val_sites (test_sites) is specified, the validation set (test set) will contain only the val_sites (test_sites. When False and val_sites (test_sites) is specified, the validation set (test set) will also contain reaches in the training set."

jds485 replied:
And yes, I didn't program the code to enforce a strict partition. I could do that, or add a warning when val_sites (test_sites) is specified and strict_spatial_partition is False

janetrbarclay replied:
I think that revised description is good. The only change I'd suggest is that the validation set (test set) "may" also contain  . . . (in the second line)

I think calling the variable "strict_spatial_partition" suggests there will be no overlap. So maybe add something to enforce that (if that's the functionality you want) or maybe change the name to something like "explicit_spatial_partition" (since the current code additions allow you to explicitly specify which reaches are in which partition)

jds485 replied:
> The only change I'd suggest is that the validation set (test set) "may" also contain

I think it's guaranteed to contain the training reaches based on this code. For example, in Line 685-687, if test_sites are specified, they are removed from training and validation, but there's no removal of training and validation sites from y_tst.

jds485 replied:
Sure, I can change to "explicit_spatial_partition"

janetrbarclay replied:
> > The only change I'd suggest is that the validation set (test set) "may" also contain
> 
> I think it's guaranteed to contain the training reaches based on this code. For example, in Line 685-687, if test_sites are specified, they are removed from training and validation, but there's no removal of training and validation sites from y_tst.

I think my thought was that the reaches could be functionally excluded due to lack of observations (especially if there is also a temporal partitioning), but maybe that's not worth clarifying

jds485 replied:
oh, okay, that makes sense. How about changing to something like:
"When False and val_sites (test_sites) is specified, the spatial removal of data from the validation set (test set) will not remove reaches that are in the training set."

jds485 commented on river_dl/evaluate.py:
```diff
@@ -268,6 +333,54 @@ def partition_metrics(
+            .apply(calc_metrics)
+            .reset_index()
+            )
+        elif group == "biweekly":
```
> I'm not sure how monthly and yearly are different from ['seg_id_nat','month'] and ['seg_id_nat','year'] since they are both grouping on the same things

The biweekly, monthly and yearly options are resampling the daily timeseries to those time steps by taking the sum of the data within those time periods (only for the days with observations). I'm not sure that sum is the best option and am open to other suggestions.

jds485 replied:
I agree with removing the "seg_id_nat" comparison to be more generic, but that will affect snakemake workflows. For example, the workflow examples all have a function that defines `group` using seg_id_nat. Might be better to address this problem in a separate issue

jds485 replied:
> So are you summing by reach and then calculating metrics across all the reaches?

yes, that's right

jds485 replied:
> I can add the group option that reports these metrics by reach

I added for reach-monthly and reach-yearly timeseries. reach-biweekly will require an additional function to define a biweekly index because Python datetime objects do not have a default biweekly index. I can add an enhancement issue for that

jds485 commented on river_dl/evaluate.py:
```diff
@@ -136,7 +152,7 @@ def calc_metrics(df):
     pred = df["pred"].values
     obs, pred = filter_nan_preds(obs, pred)
 
-    if len(obs) > 10:
```
I can revert back to 10 for now and we can convert this to an issue. 

Some ideas that could go into a new issue on this:
We could make this a function parameter.
Also, could add a count() variable to the metrics tables to know the maximum number of observations used to compute the metrics in each row.

jds485 commented:
Thanks for the review, @janetrbarclay! I think I addressed all of your comments. I think a few will require feedback from others before editing / reverting

janetrbarclay commented on river_dl/evaluate.py:
```diff
@@ -224,27 +237,79 @@ def partition_metrics(
-            data = data[~data[spatial_idx_name].isin(train_sites)]
-        if val_sites and partition=='tst':
-            data = data[~data[spatial_idx_name].isin(val_sites)]
+        if train_sites and partition == 'trn':
```
Sounds good.

jsadler2 replied:
Oh wow. Yeah. This was definitely a bug! 😮 Thanks for catching this.

jds485 replied:
@janetrbarclay: in a [comment below](https://github.com/USGS-R/river-dl/pull/211/files#r1139385726), Jeff suggests we assume that sites within the train/val/test are the only sites in those partitions. That's also what I would expect. Do you know of anyone who is relying on the previous method wherein sites that are not within train_sites/val_sites/test_sites could be in the train/val/test partitions?

janetrbarclay commented:
> Thanks for the review, @janetrbarclay! I think I addressed all of your comments. I think a few will require feedback from others before editing / reverting

Agreed! Let me know when you want me to glance at those parts again and I'll be happy to approve them.

jsadler2 commented on river_dl/evaluate.py:
```diff
@@ -162,12 +178,11 @@ def calc_metrics(df):
-            "nse_top10": percentile_metric(obs, pred, nse_eval, 90, less_than=False),
-            "nse_bot10": percentile_metric(obs, pred, nse_eval, 10, less_than=True),
-            "nse_logged": nse_logged(obs, pred),
+            "kge_logged": kge_logged(obs, pred),
```
It would be nice if you could specify which metrics you want instead of having them hard-coded. There's probably a clever way of doing that and exposing that to the top level "combined_metrics" function.

jds485 replied:
It's very quick to compute all of these metrics. I'd suggest this as an enhancement issue

jsadler2 commented on river_dl/evaluate.py:
```diff
@@ -224,27 +239,79 @@ def partition_metrics(
 
     for data_var, data in var_data.items():
+        #multiindex df
+        data_multiind = data.copy(deep=True)
```
After reading over this code, I'm not convinced that we need a `data_multiind`. Can you explain why you are adding it, @jds485 ?

jds485 replied:
Addressed in the most recent [commit](https://github.com/USGS-R/river-dl/pull/211/commits/89ec644c9f32837930c42013573924329caa87ea). I'm using only the multi-index dataframe and called it `data`

jsadler2 commented on river_dl/evaluate.py:
```diff
@@ -224,27 +239,79 @@ def partition_metrics(
-            data = data[~data[spatial_idx_name].isin(train_sites)]
-        if val_sites and partition=='tst':
-            data = data[~data[spatial_idx_name].isin(val_sites)]
+        if train_sites and partition == 'trn':
+            # simply use the train sites when specified.
+            data = data[data[spatial_idx_name].isin(train_sites)]
+            data_multiind = data_multiind.loc[data_multiind
+                                              .index
+                                              .get_level_values(level=spatial_idx_name)
+                                              .isin(train_sites)]
+        else:
+            #check if validation or testing sites are specified
+            if val_sites and partition == 'trn':
+                data = data[~data[spatial_idx_name].isin(val_sites)]
+                data_multiind = data_multiind.loc[~data_multiind
+                                                  .index
+                                                  .get_level_values(level=spatial_idx_name)
+                                                  .isin(val_sites)]
+            if test_sites and partition == 'trn':
+                data = data[~data[spatial_idx_name].isin(test_sites)]
+                data_multiind = data_multiind.loc[~data_multiind
+                                                  .index
+                                                  .get_level_values(level=spatial_idx_name)
+                                                  .isin(test_sites)]
```
I think it would be more readable to have another layer of conditions with the top just checking the `partition`  (somehow my spacing got off, so DO NOT COMMMIT AS IS :))

```suggestion
        if partition == 'trn':
            if train_sites:
                # simply use the train sites when specified.
                data = data[data[spatial_idx_name].isin(train_sites)]
                data_multiind = data_multiind.loc[data_multiind
                                                  .index
                                                  .get_level_values(level=spatial_idx_name)
                                                  .isin(train_sites)]
             else:
                #check if validation or testing sites are specified
                if val_sites:
                    data = data[~data[spatial_idx_name].isin(val_sites)]
                    data_multiind = data_multiind.loc[~data_multiind
                                                      .index
                                                      .get_level_values(level=spatial_idx_name)
                                                      .isin(val_sites)]
                if test_sites:
                    data = data[~data[spatial_idx_name].isin(test_sites)]
                    data_multiind = data_multiind.loc[~data_multiind
                                                      .index
                                                      .get_level_values(level=spatial_idx_name)
                                                      .isin(test_sites)]
```

jds485 replied:
I addressed this in a recent commit

jsadler2 commented on river_dl/evaluate.py:
```diff
@@ -268,6 +335,99 @@ def partition_metrics(
             .apply(calc_metrics)
             .reset_index()
             )
+        elif group == "year":
```
This is getting to be quite verbose. I suggest we split the `group` argument into two, maybe (`group_spatially`, and `group_temporally`). The `group_spatially` would be just a boolean. The `group_temporally` would be a `str`.

then the function could be something like:
```python
if not group_spatially and not group_temporally:
    metrics = calc_metrics(data)
    # need to convert to dataframe and transpose so it looks like the
    # others
    metrics = pd.DataFrame(metrics).T
elif group_spatially and not group_temporally:
    metrics = data.groupby(spatial_idx_name).apply(calc_metrics).reset_index()
elif not group_spatially and group_temporally:
    metrics = data.groupby(pd.Grouper(index=time_idx_name, freq=group_temporally)).apply(calc_metrics).reset_index()
elif group_spatially and group_temporally:
    metrics = data.groupby([pd.Grouper(index=time_idx_name, freq=group_temporally),
                            pd.Grouper(index=spatial_idx_name)]
                            ).apply(calc_metrics).reset_index()
```

I think that should work. We'd just have to document how the `group_temporally` argument needs to work.

jsadler2 replied:
Definitely a bigger change, but I think it would be worth trying. It would also require propagating the change all the way up including any Snakefiles that are using this function.

jsadler2 replied:
Would also resolve #212

jds485 replied:
Thanks, that is much cleaner. I think one more argument would be needed for how to do temporal aggregation. I think what you've programmed would compute metrics for native timestep only (daily). I used a sum of the daily data to get biweekly, monthly, and yearly timesteps and compute metrics for those. Let me try out this edit because it will make the code much cleaner

jds485 replied:
I addressed this change in the most recent [commit](https://github.com/USGS-R/river-dl/pull/211/commits/89ec644c9f32837930c42013573924329caa87ea). I needed 4 group args to replicate the previous functionality. I think it's more generic now to using different timesteps of aggregation, but it's more difficult to understand how to apply the 4 args to compute the desired metrics. For reference, here is the function that I used in the snakemake file to assign the 4 new args to this function to compute different metrics for different groups. See the description of the 4 args in the function docstring.
```
#Order in the list is: 
#group_spatially (bool), group_temporally (False or timestep to use), sum_aggregation (bool), site_based (bool)
def get_grp_arg(wildcards):
	if wildcards.metric_type == 'overall':
		return [False, False, False, False]
	elif wildcards.metric_type == 'month':
		return [False, 'M', False, False]
	elif wildcards.metric_type == 'reach':
		return [True, False, False, False]
	elif wildcards.metric_type == 'month_reach':
		return [True, 'M', False, False]
	elif wildcards.metric_type == 'monthly_site_based':
		return [False, 'M', True, True]
	elif wildcards.metric_type == 'monthly_all_sites':
		return [False, 'M', True, False]
	elif wildcards.metric_type == 'monthly_reach':
		return [True, 'M', True, False]
```

jsadler2 commented on river_dl/evaluate.py:
```diff
@@ -268,6 +335,99 @@ def partition_metrics(
+            )
+            metrics = calc_metrics(data_calc)
+            metrics = pd.DataFrame(metrics).T
+        elif group == ["seg_id_nat", "monthly"]:
+            #filter the data to remove nans before computing the sum
+            #so that the same days are being summed in the year.
+            data_calc = (data_multiind.dropna()
+            .groupby(
+            [pd.Grouper(level=time_idx_name, freq='M'),
+             pd.Grouper(level=spatial_idx_name)])
+            .sum()
+            )
+            data_calc = data_calc.reset_index()
+            metrics = (data_calc
+            .groupby(spatial_idx_name)
+            .apply(calc_metrics)
+            .reset_index()
+            )
```
Not sure why you would need to break it into two steps. Did you try this (below) and it didn't work?

```suggestion
        elif group == ["seg_id_nat", "monthly"]:
            #filter the data to remove nans before computing the sum
            #so that the same days are being summed in the year.
            data_calc = (data_multiind.dropna()
            .groupby(
            [pd.Grouper(level=time_idx_name, freq='M'),
             pd.Grouper(level=spatial_idx_name)])
            .apply(calc_metrics)
            .reset_index()
            )
```

jds485 replied:
No additional `.` functions were allowed after `.sum()`, so I broke it into several lines. 

Also, noting that the original and suggested code compute metrics for different groups. The original code creates a monthly timeseries for each reach and computes reach-based metrics using the monthly timeseries. The suggested code uses data in the native timestep (e.g., daily) to compute metrics for each month-reach. Both of these are options in the latest [commit](https://github.com/USGS-R/river-dl/pull/211/commits/89ec644c9f32837930c42013573924329caa87ea)

jsadler2 replied:
I guess my question was, why do you need to do `.sum()` in the first place?

jsadler2 replied:
Okay. A little slow, but I think I got it now. So it's a two-step process because first you are aggregating by the time-space dimensions (e.g., site-biweekly) and then you are aggregating by the space (e.g., site). So you end up with just one number per site, but it's looking at bigger windows of time at a time (e.g., two weeks instead of every day). 

Am I getting that right?

jsadler2 replied:
If that's right, why did you choose `sum` instead of `mean`? It seems like `mean` would be better.

For example, if I do a rmse on the biweekly sums, I get
```
site_id
01466500      8.862246
01472104     23.094643
01472119     23.529100
014721254    20.738631
014721259    16.497109
01473499     17.143101
01473500     17.544453
01473675     17.654814
01473780     15.881052
01474500     12.225832
01480617     35.625263
01480870     36.393360
01481000     24.799420
01481500     26.861900
```

which is way high.

If I do it on the means:
```
site_id
01466500     0.633631
01472104     1.651097
01472119     1.682208
014721254    1.481826
014721259    1.179091
01473499     1.225882
01473500     1.254535
01473675     1.262302
01473780     1.136314
01474500     0.874386
01480617     2.546138
01480870     2.601188
01481000     1.772298
01481500     1.919950
```
A lot lower.

Is that what you'd expect?

jds485 replied:
> So it's a two-step process because first you are aggregating by the time-space dimensions (e.g., site-biweekly) and then you are aggregating by the space (e.g., site). So you end up with just one number per site, but it's looking at bigger windows of time at a time

Yes, that's right.

> why did you choose sum instead of mean? 

Good point - the daily values are daily averages, not the total flow in a day. For some reason I was thinking of total flow, salinity, etc. I think the sum metrics will also be inflated compared to the mean. Here's an example for biweekly metrics:

```
metrics_sum
Out[126]: 
rmse               10407.172144
nse                    0.946307
rmse_top10         28507.780134
rmse_bot10          1128.871582
rmse_logged            0.162555
mean_bias          -3456.110397
mean_bias_top10   -11188.212460
mean_bias_bot10     -618.612363
nse_top10              0.793622
nse_bot10              0.803559
nse_logged             0.963944
kge                    0.899966
kge_logged             0.977389
kge_top10              0.881643
kge_bot10              0.875451
n_obs                966.000000
dtype: float64
```
```
metrics_mean
Out[127]: 
rmse                47.075655
nse                  0.649618
rmse_top10         121.434744
rmse_bot10          21.972157
rmse_logged          0.162555
mean_bias          -21.274822
mean_bias_top10    -88.179638
mean_bias_bot10    -15.051474
nse_top10           -1.496829
nse_bot10            0.258352
nse_logged           0.698174
kge                  0.756805
kge_logged           0.887974
kge_top10            0.220643
kge_bot10            0.732286
n_obs              966.000000
dtype: float64
```

jsadler2 commented on river_dl/preproc_utils.py:
```diff
@@ -674,10 +680,15 @@ def prep_y_data(
     if val_sites:
         y_trn = y_trn.where(~y_trn[spatial_idx_name].isin(val_sites))
+        y_tst = y_tst.where(~y_tst[spatial_idx_name].isin(val_sites))
+        if explicit_spatial_partition:
+            y_val = y_val.where(y_val[spatial_idx_name].isin(val_sites))
```
```suggestion
        y_val = y_val.where(y_val[spatial_idx_name].isin(val_sites))
```

jsadler2 commented on river_dl/preproc_utils.py:
```diff
@@ -674,10 +680,15 @@ def prep_y_data(
     if test_sites:
         y_trn = y_trn.where(~y_trn[spatial_idx_name].isin(test_sites))
         y_val = y_val.where(~y_val[spatial_idx_name].isin(test_sites))
+        if explicit_spatial_partition:
+            y_tst = y_tst.where(y_tst[spatial_idx_name].isin(test_sites))
```
```suggestion
        y_tst = y_tst.where(y_tst[spatial_idx_name].isin(test_sites))
```

jsadler2 commented on river_dl/preproc_utils.py:
```diff
@@ -596,6 +596,7 @@ def prep_y_data(
     test_end_date=None,
     val_sites=None,
     test_sites=None,
+    explicit_spatial_partition=False,
```
I tend to think that we should just assume that it is explicit. If I specify train/val/test sites, I expect those to be the only sites in the partition.

jds485 replied:
I agree. I added this argument to allow for the previous implementation. It sounds like the previous implementation was not intentional, so I'm happy to remove the `explicit_spatial_partition` arg here

jsadler2 replied:
Okay. I just realized something when I was working on the DO project that may change this conversation.

What if you have sites that are in the training set for the training time period, but have data in validation time period.

As the code is currently written, the data at training sites but in the validation time period will be included in the validation set. 

With these suggested changes, even if a training site has data in the validation time period, that data will not be used in the evaluation.

I think that's why I wrote the code as it is.

janetrbarclay replied:
Jumping back into this after not thinking about it for awhile, I'm wanting to make sure I have a clear understanding in my head of the options. 

It seems we can have
1. `trn_sites` (and `val_sites` and `tst_sites`) as the only sites for the respective partitions (ie remove all other sites from those partitions, but those sites can appear in other partitions), _or_ 
2.  `trn_sites` (and `val_sites` and `tst_sites`) can appear only in those partitions (ie remove those sites from all other partitions, but other sites can appear in those partitions), _or_
3.  `trn_sites` (and `val_sites` and `tst_sites`) are the only sites for those partitions and they only appear there (ie remove those sites from other partitions and remove other sites from those partitions)

Right?

I'm not quite sure which option `explicit_spatial_partition`==TRUE or FALSE maps to.

Also, would we ever want different behavior among the partitions? (ie `trn_sites` can appear in the test partition, but `tst_sites` can't appear in the training partition) It would get messy and complicated and likely isn't worth adding unless there's a clear need.

I like the idea of a check to make sure that data are not used in multiple partitions. Maybe that's an addition that is irrespective of the partition approach that's used? (just a final step in prepping the data)

jds485 replied:
>  I'm wanting to make sure I have a clear understanding in my head of the options.

Yes, I think these are the 3 options.

> I'm not quite sure which option explicit_spatial_partition==TRUE or FALSE maps to.

option 3 is explicit_spatial_partition==TRUE 
- explicit_spatial_partition==FALSE currently allows training sites to appear in validation and testing partitions. It's otherwise the same as option 3.
option 1 would be the same as option 3 if all 3 partitions were specified (right?).,
Prior to this PR, option 2 was used for tst_sites and partially used for val_sites (val sites removed from training but not from testing)

> Also, would we ever want different behavior among the partitions? (ie trn_sites can appear in the test partition, but tst_sites can't appear in the training partition) It would get messy and complicated and likely isn't worth adding unless there's a clear need.

I think a preference existed prior to this PR. The function could be modified to allow for specifying which partitions should have explicit spatial partitioning. But I agree that could be complicated. I don't currently have a need for that feature.

> I like the idea of a check to make sure that data are not used in multiple partitions. Maybe that's an addition that is irrespective of the partition approach that's used? (just a final step in prepping the data)

Yes, I agree this should happen regardless of the other choices made.

janetrbarclay replied:
> option 1 would be the same as option 3 if all 3 partitions were specified (right?).,
yes, that's right (though I wasn't thinking of that yesterday). If both of these options remain, maybe a note goes in the comments pointing this out?

I'm trying to think about how I would use this. My inclination is that if I'm specifying training sites, then I want those to be the only training sites. (otherwise what's the point of specifying sites?) So I guess I can see uses for options 1 and 3, but I'm not sure the point of # 2. 
 
What do you think of keeping those options, with the `explicit_spatial_partition` as the flag that determines which one?

jds485 replied:
Sounds good to me! That's also how I would expect a function to use provided site partitions.

janetrbarclay replied:
and if someone wanted something like the prior functionality of some get removed from other partitions and some don't, then they can fully specify all the partitions and list things how they want. (which means for option 3 it would be good for a specified list of sites for a partition to override removing sites specified for other paritions -> ie if `trn_sites` is specified, then those sites are removed from the validation partition, unless `val_sites` is also specified, in which case that takes precedence.

jds485 replied:
Sure, I can work that option in. I'll plan to get to this next week

janetrbarclay replied:
Thanks for cleaning this up for us!

jsadler2 reviewed (changes requested):
@jds485 - thanks for catching these things that have been slipping through the cracks! I have a few comments/suggestions that I think you/we should consider before making this merge.

jds485 commented on river_dl/evaluate.py:
```diff
@@ -214,62 +232,180 @@ def partition_metrics(
+    in the calendar year are used to compute a metric for that year). Note that
+    for month, 'M', this would normally have groups by year-month. We have forced
+    the groups to the 12 calendar months instead.
+    :param site_based [bool] Only applies when group_spatially is False,
```
One edit that may help is stating the number of observations used. When site_based = True, the observations are each group_temporally timestep for each reach with observations (max of n_segs*n_timesteps). When site_based = False, the observations are each group_temporally timestep with observations (max of n_timesteps).

jds485 commented on river_dl/evaluate.py:
```diff
@@ -214,62 +233,191 @@ def partition_metrics(
-            data = data[~data[spatial_idx_name].isin(val_sites)]
-
-        if not group:
+        if partition == 'trn':
```
@janetrbarclay here is the new handling of spatial data holdouts. It's similar to the preproc_utils, except this is first checking for partition of the data.

janetrbarclay replied:
Looks good to me. Thanks for doing this!

jds485 commented on river_dl/preproc_utils.py:
```diff
@@ -671,14 +677,27 @@ def prep_y_data(
 
 
-    # replace validation sites' (and test sites') data with np.nan
+    # replace trn, val and tst sites' data with np.nan
```
@janetrbarclay here is the new handling of spatial data holdouts in preproc_utils

janetrbarclay replied:
Looks good too.

janetrbarclay commented on river_dl/preproc_utils.py:
```diff
@@ -1150,3 +1187,32 @@ def read_exclude_segs_file(exclude_file):
         d = yaml.safe_load(s)
     return [val for key, val in d.items()]
+
+def check_partitions(data):
```
Yes, though I hadn't thought through the implementation details, this is the functionality I was picturing. My only suggestion has to do with pretraining. I think we're not as worried about overlapping partitions in the pretraining data. It seems like good practice to have this check as the default, but maybe either add an option to not check the pretraining data and/ or to print a warning rather than fail if there are overlaps?

jds485 replied:
I can change the hard-coded True/False to a user-specified value. 
But why would overlapping partitions be okay for pretraining?

janetrbarclay replied:
I would definitely keep it as the default to check, but maybe allow an override? I'm not sure when it would be used, and maybe I'm conflating partitioning and holdouts in my head. I know that for the gw project, we only kept our holdout reaches out of the fine-tuning training, we let them be in the pretraining. And actually, river-dl doesn't currently do anything with the validation and testing partitions of the pretraining data. (that's not a very clear answer :))

jds485 replied:
That holdout description makes sense to me. You're basically building an emulator of a process model for the study region, and I don't think it's cheating to use all reaches in that step because a process model can be run anywhere that has the necessary input data. However, it would be important to remove the real observations for holdout reaches when training the process model. Otherwise, some information from those observations would inform weights and biases of the ML model.

> river-dl doesn't currently do anything with the validation and testing partitions of the pretraining data

It looks like the model training functions could use validation or testing data (e.g., for optimizing hyperparameters during pretraining). Is your statement more about how pretraining has been implemented in workflows?

Anyway, okay, I'll write in an argument that can override the pretraining check

jds485 replied:
@janetrbarclay I added a function argument that allows overriding the pretraining check. Let me know if this is good to merge.

janetrbarclay reviewed (approved):
Looks good to me. Thanks for all your work on this! (and your patience with our slow responses)

jsadler2 reviewed (approved):
Thanks for doing this, Jared.

jds485 commented:
I added 2 final commits based on discussion in #218. I am going to merge this because the commits have been reviewed there.
